provider/pkg/provider/previewOutputs.go

```diff
@@ -0,0 +1,211 @@
+func previewResourceOutputs(
+	inputs resource.PropertyMap,
+	types map[string]metadata.CloudAPIType,
+	outputs map[string]schema.PropertySpec,
+	readOnly []string,
+) resource.PropertyMap {
+	result := resource.PropertyMap{}
+	for name, prop := range outputs {
+		key := resource.PropertyKey(name)
+		if inputValue, ok := inputs[key]; ok {
+			result[key] = previewOutputValue(inputValue, types, &prop.TypeSpec, filterAndReturnNested(readOnly, name))
+			// otherwise we could have one of two cases
+			// 1. the property is an input & output property, but does not have an input value
+			// 2. the property is a readonly (output only) property
+			// In either case, the property could be computed so we should default to marking it as computed
+		} else if isReadOnly(readOnly, name) {
+			result[key] = resource.MakeComputed(resource.NewStringProperty(""))
+		}
+	}
+	return result
+}
+
+// populateStableOutputs updates the preview outputs with the outputs from
+// the prior state, but only in certain cases.
+//
+// There is no way (based on the schema) to tell which output properties are stable
+// (i.e. they will only change if the resource is replaced). Because of this we
+// cannot blanket copy all outputs from the prior state. The ones that are not stable
+// should remain computed since they could be changing.
+//
+// Instead, we use some heuristics to determine which outputs are stable and which are not.
+// We are making the assumption that Name, Id, and Arn are stable outputs. Sometimes resources
+// will have `name`, `id`, and `arn` properties, and sometimes they will have `resourceName`, `resourceId`,
+// and `resourceArn` properties.
+func populateStableOutputs(
+	outputsFromInputs resource.PropertyMap,
+	outputsFromPriorState resource.PropertyMap,
+	readOnly []string,
+	resourceTypeName tokens.TypeName,
+) resource.PropertyMap {
+	for _, readOnlyProp := range readOnly {
+		// For now skip outputs that are part of an array property.
+		// We have no way of knowing which item in the outputs array corresponds
+		// to which item from the inputs array (since inputs could be changing).
+		if strings.Contains(readOnlyProp, "/*/") {
+			continue
+		}
+		if isStableOutput(readOnlyProp, resourceTypeName) {
+			// nested property. Sometimes the resource Arn is in a nested property
+			if strings.Contains(readOnlyProp, "/") {
+				props := strings.Split(readOnlyProp, "/")
+				current := props[0]
+				key := resource.PropertyKey(current)
+				if outputFromInput, ok := outputsFromInputs[key]; ok {
+					if outputFromState, ok := outputsFromPriorState[key]; ok {
+						outputsFromInputs[key] = resource.NewObjectProperty(populateStableOutputs(
+							outputFromInput.ObjectValue(),
+							outputFromState.ObjectValue(),
+							[]string{strings.Join(props[1:], "/")},
+							resourceTypeName,
+						))
+					}
+				}
+			} else {
+				key := resource.PropertyKey(readOnlyProp)
+				if output, ok := outputsFromPriorState[key]; ok {
```

Comment:
This branch will not consider outpusFromInputs? Why?

Reply:
`outputsFromInputs` will have all possible outputs properties because of `previewResourceOutputs`, so if it's a stable property and it exists in `outputsFromPriorState` then copy it over.